rdtools/degradation.py

```diff
@@ -74,82 +71,107 @@ def degradation_with_ols(normalized_energy):
-    if isinstance(dataframe.index, pd.DatetimeIndex):
-        dataframe = dataframe.resample('MS').mean()
-        
-        if interpolate_flag:
-	    dataframe = dataframe.interpolate()
-        else:
-            # append the median value to missing months
-            dataframe = dataframe.fillna(value = energy_median)
+    df_check_freq = df_check_freq.dropna()
 
-    # assume an array of months
-    dataframe['Month'] = np.arange(0, len(dataframe))
+    if df_check_freq.index.freq is None:
+        raise ValueError('Classical decomposition requires a regular time series with'
+                         ' defined frequency and no missing data.')
 
-    if pd.infer_freq(dataframe.index) == 'MS' and len(normalized_energy) > 12:
-        dataframe = dataframe.rolling(window = 12, center = True).mean()
-        
-    dataframe = dataframe.dropna()
-    y2 = dataframe['normalized_energy']
-    x2 = dataframe['Month']
+    #calculate a years column as x value for regression, ignoreing leap years
+    day_diffs = (df.index - df.index[0])
+    df['days'] = day_diffs.astype('timedelta64[s]')/(60*60*24)
+    df['years'] = df.days/365.0
     
+    #Compute yearly rolling mean to isolate trend component using moving average
+    it = df.iterrows()
+    energy_ma = []
+    for i, row in it:
+        if row.years-0.5 >= min(df.years) and row.years+0.5 <= max(df.years):
+            roll = df[(df.years <= row.years+0.5) & (df.years >= row.years-0.5)]
+            energy_ma.append(roll.normalized_energy.mean())
+        else:
+            energy_ma.append(np.nan)
+    
+    df['energy_ma'] = energy_ma
+    
+    #add intercept-constant to the exogeneous variable
+    df = sm.add_constant(df)
+    
+    #perform regression
+    ols_model = sm.OLS(endog = df.energy_ma, exog = df.loc[:,['const','years']],
+                       hasconst = True, missing = 'drop' )
+    
+    results = ols_model.fit()
+    
+    # collect intercept and slope
+    b, m = results.params
+    
+    # rate of degradation in terms of percent/year
+    Rd_pct = 100.0 * m / b
+    
+    #Calculate RMSE
+    rmse = np.sqrt(results.mse_resid)
+    
+    #Collect standrd errors
+    stderr_b, stderr_m = results.bse
 
+    #Perform Mann-Kendall 
+    test_trend, h, p, z = _mk_test(df.energy_ma.dropna(), alpha=0.05)
 
-    if len(dataframe) <= 2:
-        print '\nNot enough data for seasonal decomposition:'
-        return {}
-        
-    # OLS regression with constant
-    x2 = sm.add_constant(x2)
-    model2 = sm.OLS(y2,x2).fit()
-    b_cd, m_cd = model2.params
-    Rd_cd, SE_Rd_cd = ols_rd_uncertainty(model2)    
-
-    print '\nDegradation and Standard Error of Classical decomposition:'
-    print 'Rd = {:.2f} +/- {:.2f}'.format(Rd_cd, SE_Rd_cd)  
+    #Monte Carlo for error in degradation rate
+    dist = [z[1]/z[0] for z in np.random.multivariate_normal(results.params, results.cov_params(), 10000)]
+    Rd_CI = np.percentile(dist, [50-34.1, 50+34.1])*100.0
```

Comment:
Since the confidence interval logic is used in the `degradation_ols` function as well, perhaps it should be factored-out into a separate function. That way, if there are changes to this logic, it only needs to be changed in one place.